lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs

```diff
@@ -526,29 +526,119 @@ balanceTransaction
     pp
     timeTranslation
     utxoAssumptions
-    utxo
+    UTxOIndex {availableUTxO, availableUTxOIndex}
     genChange
     s
-    partialTx
+    PartialTx {extraUTxO, tx, redeemers, timelockKeyWitnessCounts}
     = do
-    let adjustedPartialTx = flip (over #tx) partialTx $
-            assignMinimalAdaQuantitiesToOutputsWithoutAda pp
+    guardExistingCollateral
+    guardExistingReturnCollateral
+    guardExistingTotalCollateral
```

Comment:
🤔 I wonder if it doesn't still make more sense to keep these in the inner function — close to the main logic and post conditions...

Reply:
> 🤔 I wonder if it doesn't still make more sense to keep these in the inner function — close to the main logic and post conditions...

In `master`, these conditions are potentially checked **_twice_**:
1. When called with `SelectionStrategyOptimal`
2.  When called with `SelectionStrategyMinimal`.
However the outcome will be identical in both cases, as these checks are purely a function of the input to `balanceTx`.

So I think there's no real point to checking these conditions twice.

Furthermore, since the inner function is not exported outside of the `Balance` module, I think it's safe to move these checks to the outer function.

Reply:
@Anviking As discussed, I've temporarily moved the collateral-related guards back to the inner function in c7983a75a2738ead33d21e43a713f8f91cc4ed11.

Reply:
I checked in with Heinrich for a <s>second</s>third opinion who didn't see a conceptual problem with moving the guards further from the inner function where they are needed. So sure — let's go with that.

Also, it might be that we're not too far away from making `balanceTransactionInner` much _smaller_ than `balanceTransaction`, as opposed to the other way around we have on master, which I believe also would address my concern (of having two fairly complex entangled functions).

Reply:
@Anviking Fair enough. I've removed c7983a75, restoring the PR back to its original state. Thanks!

Reply:
> Also, it might be that we're not too far away from making balanceTransactionInner much smaller than balanceTransaction

Although I'm not sure how this will turn out 🤔 

Couldn't what you're doing in this PR be contained in a helper function like this:
```haskell
combineUTxOEnv
    :: PartialTx era 
    -> UTxOIndex era
    -> Either ErrInconsistentUTxOs (UTxO era, UTxOSelection era) -- (utxoReference, utxoSelection)
```
which could be called by either `balanceTransaction` or `balanceTransactionInner` (modulo `where` helpers depending on args) depending on how we split things?